src/hint_deploy.py

```diff
@@ -126,6 +129,25 @@ def hint_constellation(cfg):
     return obj
 
 
+def hint_upgrade_hintr(obj):
+    hintr = obj.containers.find("hintr")
+    worker = obj.containers.find("worker")
+    container = hintr.get(obj.prefix)
+
+    # Always pull the docker image - and do this *before* we start
+    # removing things to minimise downtime.
+    docker_util.image_pull(hintr.name, str(obj.data.hintr_ref))
+    docker_util.image_pull(hintr.name, str(obj.data.hintr_worker_ref))
+
+    if container:
+        if container.status == "running":
+            print("Stopping previous hintr and workers")
+            container.exec_run(["hintr_stop"])
+        docker_util.container_remove_wait(container)
+
+    obj.start(subset=[hintr.name, worker.name])
```

Comment:
So, to be clear, we're not removing the existing workers at all, just adding more new ones? But the hintr stop command prevents the outdated ones from accepting jobs? Is there some cleanup mechanism for the old workers?

Reply:
The old workers will exit when they finish their last job, but will kick around in exited state. Docker has a server-side AutoRemove option which would help, but that has caused race conditions elsewhere and so I've not used it. We should arrange to periodically run `docker container prune` on the server which will clean out the old exited containers